File: apps/backend-go/internal/service/auth_service.go
package service

import (
	"context"
	"errors"

	"golang.org/x/crypto/bcrypt"

	"github.com/xiaoshuai1024/luban-backend-go/internal/model"
	"github.com/xiaoshuai1024/luban-backend-go/internal/repository"
)

var (
	ErrInvalidCredentials = errors.New("INVALID_CREDENTIALS")
	ErrUserDisabled       = errors.New("USER_DISABLED")
)

type AuthService struct {
	userRepo *repository.UserRepository
}

func NewAuthService(userRepo *repository.UserRepository) *AuthService {
	return &AuthService{userRepo: userRepo}
}

type LoginResult struct {
	User   *model.User       `json:"user"`
	Claims map[string]string `json:"claims"`
}

func (s *AuthService) Login(ctx context.Context, username, password string) (*LoginResult, error) {
	u, err := s.userRepo.FindByUsername(ctx, username)
	if err != nil {
		if errors.Is(err, repository.ErrUserNotFound) {
			return nil, ErrInvalidCredentials
		}
		return nil, err
	}
	if u.Status != "active" {
		return nil, ErrUserDisabled
	}
	if bcrypt.CompareHashAndPassword([]byte(u.Password), []byte(password)) != nil {
		return nil, ErrInvalidCredentials
	}
	return &LoginResult{
		User: u,
		Claims: map[string]string{
			"userId": u.ID,
			"role":   u.Role,
		},
	}, nil
}

// Me 按 userId 返回当前登录用户的完整信息（用于 /auth/me）。
//
// 对齐 Java AuthService.me：从 DB 按 id 取最新 user 并返回。User.Password 字段
// 标了 json:"-"，序列化时自动脱敏，不会把密码哈希暴露给前端。
// 用户不存在返回 ErrUserNotFound（对齐 Java BusinessException.userNotFound()）。
func (s *AuthService) Me(ctx context.Context, userID string) (*model.User, error) {
	return s.userRepo.Get(ctx, userID)
}

File: apps/backend-java/src/main/java/com/luban/backend/service/AuthService.java
package com.luban.backend.service;

import com.luban.backend.dto.LoginResponse;
import com.luban.backend.dto.UserResponse;
import com.luban.backend.entity.User;
import com.luban.backend.exception.BusinessException;
import com.luban.backend.mapper.UserMapper;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Service
public class AuthService {

    private final UserMapper userMapper;
    private final PasswordEncoder passwordEncoder;

    public AuthService(UserMapper userMapper, PasswordEncoder passwordEncoder) {
        this.userMapper = userMapper;
        this.passwordEncoder = passwordEncoder;
    }

    public LoginResponse login(String username, String password) {
        User u = userMapper.findByUsername(username);
        // 先统一校验密码（用户不存在/密码错误同响应，消除账号状态 oracle：
        // 错误密码不再暴露 pending_verification/disabled 等账号状态），通过后才判账号状态
        if (u == null || !passwordEncoder.matches(password, u.getPassword())) {
            throw BusinessException.invalidCredentials();
        }
        // 注册未验证邮箱：文案「邮箱未验证」引导回注册流程（plan §3.2 非法态）
        if ("pending_verification".equals(u.getStatus())) {
            throw BusinessException.userPendingVerification();
        }
        if (!"active".equals(u.getStatus())) {
            throw BusinessException.userDisabled();
        }
        return LoginResponse.of(UserResponse.fromEntity(u));
    }

    /**
     * Current user for /auth/me (by userId from UserContext). Returns full user from DB.
     */
    public UserResponse me(String userId) {
        User u = userMapper.getById(userId);
        if (u == null) {
            throw BusinessException.userNotFound();
        }
        return UserResponse.fromEntity(u);
    }
}
